Add support for alpha channel in wxGenericColourDialog.

a-wi opened the pull request:
See http://trac.wxwidgets.org/ticket/14127

a-wi committed:
Optimize wxGenericColourDialog

Variable to hold the size of the custom colour box doesn't need to be a member variable since it's used only locally in wxGenericColourDialog::CalculateMeasurements.

a-wi committed:
Fix type of member variable in wxGenericColourDialog

Variable to hold the size should be of type wxSize, not wxPoint.

tierra commented:
It doesn't look like there's any kind of new style/flag for controlling whether the user is presented with controls for setting alpha values. I believe that in a lot of (if not most) cases, applications currently using these dialogs don't actually want to give the user the option of setting a custom alpha value for their colour - even if it is perfectly possible to do so.

I feel like this should only be enabled with a new control style explicitly passed to enable it.

a-wi commented:
I agree that it would be good to control this option with special flag but AFAICS (I'm not an OSX expert) it is already implemented under OSX as an unconditional feature (http://trac.wxwidgets.org/ticket/14127#comment:2) so there is a question of inter-system consistency.

vadz commented:
Yes, I agree with @tierra too, it's misleading to allow the user to select opacity if it's just going to be ignored anyhow, so it would be better to add a flag for it.

FWIW OS X `NSColorPanel` has `showsAlpha` property which could apparently be used to implement this flag there.

tierra commented:
Hmm, that is unfortunate. I wonder if it would be worth providing the style, and just defaulting it to enabled on OSX (and just OSX) for back-compat.

a-wi commented:
For wxColourPicker opacity support could be controlled by the flag (e.g. wxCLRP_USE_OPACITY) which is passed as parameter to ctor (currently wxCLRP_USE_TEXTCTRL and wxCLRP_SHOW_LABEL flags are used this way).

Unfortunately, in the current implementation of wx(Generic)ColourDialog no flags are passed to ctor so adding opacity control could by implemented by:
1. Adding new argument with flags in ctor.
or
2. Extending wxColourData by adding new flag to control opacity support (with new accessors similar to Set/GetChooseFull - something like Set/GetUseOpacity) and use this flag in the colour dialogues to enable/disable opacity-related features.

Which option would be better?

a-wi commented:
It seems to me that opacity control done through wxColourData would be more in line with wxWidgets spirit so wxColourData has been extended with additional additional flag (and corresponding methods) to allow this control.
Opacity can be enabled/disabled in colour dialogs through wxColourData::SetShowOpacity (in analogy with wxColourData::SetChooseFull).

tierra commented:
Hmm, I was actually kind of leaning more towards the traditional dialog styles used in the dialog constructor. I know it doesn't have any styles right now, but that is the more common approach to various built-in dialog features, and feels like a legitimate reason to add the style flags now. Although, I'm not really in any position to make any kind of authoritative calls one way or another, I haven't touched this stuff in years, so I held off on mentioning anything.

vadz commented on include/wx/generic/colrdlgg.h:
```diff
@@ -14,18 +14,18 @@
 
+// Preview with opacity is possible only
+// if wxGCDC and wxStaticBitmap are available.
+#define USE_PREVIEW_WITH_ALPHA (wxUSE_GRAPHICS_CONTEXT && wxUSE_STATBMP)
```
This symbol should have a "wx" prefix as it's in a public header.

vadz commented on src/common/colourdata.cpp:
```diff
@@ -102,6 +107,17 @@ bool wxColourData::FromString(const wxString& str)
     }
+
+    token = tokenizer.GetNextToken();
+    if ( token.empty() )
```
This is very minor but why not just do `m_showOpacity = token == "1";`, it seems like it would have exactly the same effect (except it would also work with strings in invalid format, e.g. using "2").

vadz commented on src/generic/colrdlgg.cpp:
```diff
@@ -37,6 +37,19 @@
 #include "wx/generic/colrdlgg.h"
 
+#if USE_PREVIEW_WITH_ALPHA
+#include <wx/statbmp.h>
```
We should use quotes and not angle brackets around the includes in wxWidgets itself.

vadz commented on src/generic/colrdlgg.cpp:
```diff
@@ -131,6 +144,17 @@ wxGenericColourDialog::wxGenericColourDialog(wxWindow *parent,
 
 wxGenericColourDialog::~wxGenericColourDialog()
 {
+#if USE_PREVIEW_WITH_ALPHA
```
Why do we need to call `Unbind()` explicitly when the dialog is being destroyed anyhow? AFAICS this code shouldn't be necessary at all...

vadz commented on src/generic/colrdlgg.cpp:
```diff
@@ -148,6 +172,10 @@ bool wxGenericColourDialog::Create(wxWindow *parent, wxColourData *data)
         m_colourData = *data;
 
+#if !USE_PREVIEW_WITH_ALPHA
+    wxASSERT_MSG( !m_colourData.GetShowOpacity(), wxS("Unable to preview ARGB colours") );
```
I don't think it warrants an assert. If we can't allow the user to select alpha channel, we don't, it should just be mentioned in the documentation that this functionality might be not available under all platforms/in all build configurations.

vadz commented:
I can see both sides of the argument and while I'm not sure which possibility is the best one, at the very least I agree that `SetShowOpacity()` seems to be quite similar to `SetChooseFull()` and as we always had the latter, why not add the former.

Perhaps it would be worth calling it `SetChooseOpacity()` (or `SetChooseAlpha()`?) for even more consistency though?

a-wi committed:
Add opacity control feature to wxColourData.

Added internal flag (with corresponding accessors) which purpose is to enable/disable modifying the opacity of the colours in colour dialogs.

a-wi commented:
Commits modified as suggested.
Additionally added preview of true ARGB colours in the form of chessboard grid for easier distinction from RGB-only colours (idea taken from GTK+).

a-wi committed:
Add alpha channel support to wxGenericColourDialog.

Added opacity slider and custom colours (with alpha) previews (implemented with wxStaticBitmap controls). This feature is controlled by wxColourData::GetChooseAlpha().

a-wi committed:
Add slider labels in wxGenericColourDialog.

a-wi committed:
Modify dialogs sample to present also generic colour dialog with alpha support.

Added separate submenu and menu items to show generic colour dialogs with opacity control enabled/disabled.

a-wi committed:
Fix initialization of custom colours for generic colour dialogs in dialogs sample.

Use number of custom colours taken from wxColourData instead of using explicit numeric value.

a-wi committed:
Don't use directly wxColourData data members in wxGenericColourDialog.

wxColourData data members are public for backwards compatibility only and shouldn't be used directly.

a-wi committed:
Update documentation regarding alpha support in wxGenericColourDialog.
